spec/fastsync/vdd/fastsyncL2.tla

```diff
@@ -0,0 +1,528 @@
+    LET bStore == 
+    [h \in Heights |-> IF h \in failedRequests THEN NilBlock ELSE bPool.blockStore[h]] IN  \* Should we remove received block in this case?
+    
+    IF peer \in bPool.peerIds     
+    THEN [bPool EXCEPT 
+            !.peerIds = pIds,
+            !.peerHeights = pHeights,
+            !.pendingBlocks = pBlocks,
+            !.receivedBlocks = rBlocks,
+            !.blockStore = bStore
+          ]                              
+    ELSE bPool
+     
+\* If valid (and timely) status response, update peerHeights.
+\* If invalid message, then remove peer.
+\* Returns new block pool.                             
+HandleStatusResponse(msg, bPool, randomFlag) == 
+    LET peerHeight == bPool.peerHeights[msg.peerId] IN     
+    IF /\ msg.peerId \in bPool.peerIds 
+       /\ MsgOnTime(msg, randomFlag)
+       /\ msg.height >= peerHeight  
+    THEN    \* a correct and timely response 
+        LET pHeights == [bPool.peerHeights EXCEPT ![msg.peerId] = msg.height] IN
+        [bPool EXCEPT !.peerHeights = pHeights]                   
+    ELSE RemovePeer(msg.peerId, bPool)   \* the peer has sent us duplicate status message or not on time or it was not solicited response
+        
+        
+HandleBlockResponse(msg, bPool, randomFlag) == 
+    LET h == msg.block.height IN
+    IF /\ msg.peerId \in bPool.peerIds
+       /\ bPool.blockStore[h] = NilBlock 
+       /\ bPool.pendingBlocks[h] = msg.peerId 
+       /\ MsgOnTime(msg, randomFlag)  
+    THEN LET bStore == [bPool.blockStore EXCEPT ![h] = msg.block] IN 
+         LET rBlocks == [bPool.receivedBlocks EXCEPT![h] = msg.peerId] IN   
+         [bPool EXCEPT 
+            !.blockStore = bStore,
+            !.receivedBlocks = rBlocks
+         ]  
+    ELSE RemovePeer(msg.peerId, bPool)
+                       
+
+MaxPeerHeight(bPool) == 
+    IF bPool.peerIds = {}
+    THEN 0 \* no peers, just return 0
+    ELSE CHOOSE max \in { bPool.peerHeights[p] : p \in bPool.peerIds }:
+            \A p \in bPool.peerIds: bPool.peerHeights[p] <= max \* max is the maximum
+    
+
+\* returns next height for which request should be sent.
+\* returns NilHeight in case there is no height for which request can be sent.
+FindNextRequestHeight(bPool) == 
+    LET S == {i \in Heights:  
+                /\ i >= bPool.height  
+                /\ i <= MaxPeerHeight(bPool)   
+                /\ bPool.blockStore[i] = NilBlock 
+                /\ bPool.pendingBlocks[i] = NilPeer} IN
+    IF S = {} 
+        THEN NilHeight
+    ELSE      
+        CHOOSE min \in S:  \A h \in S: h >= min
+
+\* Returns number of pending requests for a given peer.
+NumOfPendingRequests(bPool, peer) == 
+    LET peerPendingRequests == 
+        {h \in Heights: 
+            /\ h >= bPool.height 
+            /\ bPool.pendingBlocks[h] = peer
+            /\ bPool.blockStore[h] = NilBlock
+        }
+    IN 
+    Cardinality(peerPendingRequests)
```

Comment:
Are you sure that you really need the cardinality here? Set cardinalities are tough for Apalache and TLAPS. Is the grouping by `PEER_MAX_REQUESTS` an implementation feature? :-)

Reply:
This was a try to model constraint that there shouldn't be more than x pending requests at a time (in addition to per peer bound). Frankly, I am not sure if this is really needed, as bounding max number per peer we achieve something similar as number of pending requests is `<= num_peers * PEER_MAX_REQUESTS`. Even in case we keep this at the implementation level, we might want removing it from the spec, for simplicity and also helping model checkers.